src/account-abstraction/actions/bundler/estimateUserOperationGas.ts

```diff
@@ -0,0 +1,153 @@
+import type { Address, Narrow } from 'abitype'
+import { parseAccount } from '../../../accounts/utils/parseAccount.js'
+import type { Client } from '../../../clients/createClient.js'
+import type { Transport } from '../../../clients/transports/createTransport.js'
+import { AccountNotFoundError } from '../../../errors/account.js'
+import type { BaseError } from '../../../errors/base.js'
+import type { ErrorType } from '../../../errors/utils.js'
+import type { Chain } from '../../../types/chain.js'
+import type { Hex } from '../../../types/misc.js'
+import type { Assign, OneOf, Prettify } from '../../../types/utils.js'
+import { getAction } from '../../../utils/getAction.js'
+import type { SmartAccount } from '../../accounts/types.js'
+import type { PaymasterActions } from '../../clients/decorators/paymaster.js'
+import type {
+  DeriveSmartAccount,
+  GetSmartAccountParameter,
+} from '../../types/account.js'
+import type {
+  DeriveEntryPointVersion,
+  EntryPointVersion,
+} from '../../types/entryPointVersion.js'
+import type {
+  EstimateUserOperationGasReturnType as EstimateUserOperationGasReturnType_,
+  UserOperation,
+  UserOperationCalls,
+  UserOperationRequest,
+} from '../../types/userOperation.js'
+import { getUserOperationError } from '../../utils/errors/getUserOperationError.js'
+import { formatUserOperationGas } from '../../utils/formatters/userOperationGas.js'
+import { formatUserOperationRequest } from '../../utils/formatters/userOperationRequest.js'
+import {
+  type PrepareUserOperationParameters,
+  prepareUserOperation,
+} from './prepareUserOperation.js'
+
+export type EstimateUserOperationGasParameters<
+  account extends SmartAccount | undefined = SmartAccount | undefined,
+  accountOverride extends SmartAccount | undefined = SmartAccount | undefined,
+  calls extends readonly unknown[] = readonly unknown[],
+  //
+  _derivedAccount extends SmartAccount | undefined = DeriveSmartAccount<
+    account,
+    accountOverride
+  >,
+  _derivedVersion extends
+    EntryPointVersion = DeriveEntryPointVersion<_derivedAccount>,
+> = Assign<
+  UserOperationRequest<_derivedVersion>,
+  OneOf<{ calls: UserOperationCalls<Narrow<calls>> } | { callData: Hex }> & {
+    paymaster?:
+      | Address
+      | true
+      | {
+          /** Retrieves paymaster-related User Operation properties to be used for sending the User Operation. */
+          getPaymasterData?: PaymasterActions['getPaymasterData'] | undefined
+          /** Retrieves paymaster-related User Operation properties to be used for gas estimation. */
+          getPaymasterStubData?:
+            | PaymasterActions['getPaymasterStubData']
+            | undefined
+        }
+      | undefined
+    /** Paymaster context to pass to `getPaymasterData` and `getPaymasterStubData` calls. */
+    paymasterContext?: unknown
```

Comment:
`undefined` gets absorbed by `unknown`, but better to have it than no imo

```suggestion
    paymasterContext?: unknown | undefined
```